images/generate_cache.py

```diff
@@ -0,0 +1,187 @@
+ @author Jonathan Thomas <jonathan@openshot.org>
+
+ @section LICENSE
+
+ Copyright (c) 2008-2018 OpenShot Studios, LLC
+ (http://www.openshotstudios.com). This file is part of
+ OpenShot Video Editor (http://www.openshot.org), an open-source project
+ dedicated to delivering high quality video editing and animation solutions
+ to the world.
+
+ OpenShot Video Editor is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ OpenShot Video Editor is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with OpenShot Library.  If not, see <http://www.gnu.org/licenses/>.
+ """
+
+import json
+import os
+from classes import info
+from classes.logger import log
+import openshot
+from PyQt5.QtWidgets import QApplication
+
+# Try to get the security-patched XML functions from defusedxml
+try:
+    from defusedxml import minidom as xml
+except ImportError:
+    from xml.dom import minidom as xml
+
+
+# Create Qt application (required for SVG font rendering)
+svg_app = QApplication([])
+icon_size = info.LIST_ICON_SIZE
+emoji_size = info.EMOJI_ICON_SIZE
+
+
+def create_cache_thumbnail(path, width, height, thumb_path):
+    """Create a thumbnail image of a specific image file at a specific size"""
+    display_scales = [1, 2]
+    mask_path = os.path.join(info.IMAGES_PATH, "mask.png")
+
+    for scale in display_scales:
+        try:
+            # Reload this reader
+            clip = openshot.Clip(path)
+            reader = clip.Reader()
+
+            # Open reader
+            reader.Open()
+
+            scale_thumb_path = thumb_path
+            scale_width = width
+            scale_height = height
+            if scale > 1:
+                # Create @2x version of cache image
+                clip.scale_x.AddPoint(1.0, 1.0 * scale)
+                clip.scale_y.AddPoint(1.0, 1.0 * scale)
+                scale_width *= scale
+                scale_height *= scale
+                suffix = "@%dx" % scale
+                thumb_path_base, thumb_path_ext = os.path.splitext(thumb_path)
+                scale_thumb_path = "%s%s%s" % (thumb_path_base, suffix, thumb_path_ext)
+
+            # Save thumbnail
+            reader.GetFrame(0).Thumbnail(scale_thumb_path, scale_width, scale_height,
+                                         mask_path, "", "#000", True, "png", 85)
+            reader.Close()
+            clip.Close()
+
+        except Exception:
+            # Handle exception
+            log.debug('Invalid cache image file %s', path, exc_info=1)
+
+
+#################################################
```

Comment:
Together with the `Image` class, those methods could be just a series of `read_images() -> List[Image]`, that would make way less procedural and easy to understand as they would be name after their semantic and not based on `#####` separating each of the parts. I have converted this one here to give an idea of what it could be:

```
def read_ui_cache() -> List[Image]:
    """TITLES UI CACHE - Create small images for UI
    get a list of files in the OpenShot package directory
    """
    titles_dir = os.path.join(info.PATH, "titles")
    titles_list = [
        os.path.join(titles_dir, filename)
        for filename in sorted(os.listdir(titles_dir))
        ]

    images = []
    for icon_path in sorted(titles_list):
        filename = os.path.basename(icon_path)
        fileBaseName = os.path.splitext(filename)[0]

        # Check for thumbnail path (in build-in cache)
        thumb_path = base_image_path / "cache" / "{}.png".format(fileBaseName)
        images.append(Image(
            path=pathlib.Path(icon_path),
            width=icon_size.width(),
            height=icon_size.height(),
            thumbnail=Thumbnail(path=thumb_path)
        ))
    return images
```

Doing that way you would at the end of the file just have a:

```
for method in [read_ui_cache, read_emojis_path, read_effects_ui_cache, read_blender_ui_cache, read_transitions_ui_cache]:
  for image in method():
     create_cache_thumbnail(image)
```
 That would make it way more fluent and clearly express its semantic.

Reply:
Thanks for the suggestions. I agree, and I'll add this onto my list for a future improvement. I'm a bit worried about changing anything at this point, since I've done a bunch of testing and validation on this PR.

Reply:
No problems @jonoomph, I can also do myself this refactoring and add a unit test.